### ISSUE TITLE
How do you get attributes from a device in a post processing function

### ISSUE THREAD

**Issue by BarlowHall** [thingsboard/thingsboard#2219]
Hello there! Hope you are having a nice day

So... Straight to the point.
I have a dashboard with a Timeseries Flot on it.
What I want is to have a variable for each device that, when modified, effects the data displayed.
E.g:
I have a timeseries flot with a random number generated every 10 seconds, which is then multiplied by a number in the post processing function. Is there anyway to have that multiplier be an attribute for each device.

Thank you in advance!

**Comment by BarlowHall** [comment]
Hello again!

I have figured out a solution!
If you create a new Timeseries Flot widget and add the function
```
self.addAttr = async function() {
    attributeService = self.ctx.$scope.$parent.$injector.get('attributeService');
    window.attr = {};
    for (var ind in self.ctx.datasources) {
        ds = self.ctx.datasources[ind];
        entity_type = ds.entityType;
        entity_id = ds.entityId;
        aliasName = ds.aliasName;
        attributes = await attributeService.getEntityAttributesValues(entity_type, entity_id, 'SHARED_SCOPE');
        
        attr[aliasName] = attributes;
    }
}
```
to it and then call the function in the self.onDataUpdated function...
Then you should be able to access the attributes from anywhere using `window.attr[AliasName]`,
where "AliasName" is the name of the entity alias you want the shared attributes from.

I haven't tested this with the alias resolved as multiple entities, but I have with two aliases.

Hope someone finds this useful!

Merry Christmas!

**Comment by nmahale10** [comment]
Hello Did you get any update on this.

**Comment by BarlowHall** [comment]
Hello! My previous post fixed my issue.
It should be as simple as putting the previously posted code into a new time series flot, and then using it in your post processing function.

**Comment by nmahale10** [comment]
Ok Thanks for the update.Actually I am trying to Get Latest Telemetry Keys on Dashboard.

If I have two devices say Device 1 and Device2.
With different  keys
Device 1 - key1 ,key2
Device 2 - key3, key4

So If I select Device1 and use “Navigate to dashboard state “ function to navigate to oder dashboard say “Dashboard1” . I am showing data on Table1 .It should show values of key and  key2.

Similarly if I select Device 2 it should show data of “key3” and “key4” on  Table1 of “Dashboard1”.


Sent from my iPhone

On 15-Jun-2020, at 9:28 PM, BarlowHall <notifications@github.com> wrote:

﻿
CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.

Hello! My previous post fixed my issue.
It should be as simple as putting the previously posted code into a new time series flot, and then using it in your post processing function.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fthingsboard%2Fthingsboard%2Fissues%2F2219%23issuecomment-644221641&data=02%7C01%7Cneha.mahale%40lntebg.com%7Cecc47762b4e14e8598b208d81144f0aa%7C4852d0fcf87a462bad09773f986ccc04%7C0%7C1%7C637278335093980616&sdata=o3ZurPMvfwiG0svIEcuZ%2FD0wLouWHy%2BGKgVAIBf1Xrg%3D&reserved=0>, or unsubscribe<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFJFEYQKDSNORANY2JKWA7LRWZAJ3ANCNFSM4JTB5VTA&data=02%7C01%7Cneha.mahale%40lntebg.com%7Cecc47762b4e14e8598b208d81144f0aa%7C4852d0fcf87a462bad09773f986ccc04%7C0%7C1%7C637278335093980616&sdata=eODZu4b7DyZktfGLJGdqq2tEn5JyEpr7TReVgBLRzDs%3D&reserved=0>.

Larsen & Toubro Limited

www.larsentoubro.com

This Email may contain confidential or privileged information for the intended recipient (s). If you are not the intended recipient, please do not use or disseminate the information, notify the sender and delete it from your system.

**Comment by BarlowHall** [comment]
Ok, this is something completely different from what my problem was.
But I can still help.

Do you already have the navigation to different dashboard states working?
And does Device1 have key3 and key4, or Device2 have key1 and key2?

And where are you trying to display it?

And since this is a different question than mine, I should suggest that you create a different issue for you question.

**Comment by nmahale10** [comment]
Ok. Thanks a lot.

I have created a setup for same in TB.

No , Device 1 does not have key3 and key4.
Similarly Device 2 does not have key1 and key2.


I am trying to display it on Table Timeseries widget.

I have already created issue for same
https://github.com/thingsboard/thingsboard/issues/2809

Got some solution as well.But unfortunately not able to implement it completely.
Sent from my iPhone

On 15-Jun-2020, at 10:00 PM, BarlowHall <notifications@github.com> wrote:

﻿
CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.

Ok, this is something completely different from what my problem was.
But I can still help.

Do you already have the navigation to different dashboard states working?
And does Device1 have key3 and key4, or Device2 have key1 and key2?

And where are you trying to display it?

And since this is a different question than mine, I should suggest that you create a different issue for you question.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fthingsboard%2Fthingsboard%2Fissues%2F2219%23issuecomment-644238881&data=02%7C01%7Cneha.mahale%40lntebg.com%7C6507ac07161547b2e38d08d811496fe8%7C4852d0fcf87a462bad09773f986ccc04%7C0%7C1%7C637278354422903801&sdata=7klCXpb82RycX6uzDdCmqA5kDbi7GpylcSIsG75Acbk%3D&reserved=0>, or unsubscribe<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFJFEYRL6NDIUO6EIRBGYMDRWZEBBANCNFSM4JTB5VTA&data=02%7C01%7Cneha.mahale%40lntebg.com%7C6507ac07161547b2e38d08d811496fe8%7C4852d0fcf87a462bad09773f986ccc04%7C0%7C1%7C637278354422903801&sdata=IxTPllQW%2F3pnEb47M4Dw%2BRyGgyL2Gwq5AVd20n3arUc%3D&reserved=0>.

Larsen & Toubro Limited

www.larsentoubro.com

This Email may contain confidential or privileged information for the intended recipient (s). If you are not the intended recipient, please do not use or disseminate the information, notify the sender and delete it from your system.

**Comment by BarlowHall** [comment]
Okay, so why doesn't it work for you? His solution seems to me like it would work.
How have you not been able to implement it completely?

**Comment by nmahale10** [comment]
I have tried adding  rule in rule engine to calculate and save something like "values" . But while saving it to Telemtry using "Save Telemetry" rule .It was giving me error .So not able to save it in telemtry.
Shared a screenshot of my script.

![values](https://user-images.githubusercontent.com/22172258/84694114-46f1d400-af66-11ea-85e0-d40be9670919.png)

**Comment by nmahale10** [comment]
I have tried adding rule in rule engine to calculate and save something like "values" . But while saving it to Telemtry using "Save Telemetry" rule .It was giving me error .So not able to save it in telemtry.
Shared a screenshot of my script.

[values]<https://user-images.githubusercontent.com/22172258/84694114-46f1d400-af66-11ea-85e0-d40be9670919.png>

Sent from my iPhone

On 15-Jun-2020, at 11:02 PM, BarlowHall <notifications@github.com> wrote:

﻿
CAUTION: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.

Okay, so why doesn't it work for you? His solution seems to me like it would work.
How have you not been able to implement it completely?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fthingsboard%2Fthingsboard%2Fissues%2F2219%23issuecomment-644269684&data=02%7C01%7Cneha.mahale%40lntebg.com%7Cd9232c8cede2473c6ece08d811521158%7C4852d0fcf87a462bad09773f986ccc04%7C0%7C1%7C637278391470561704&sdata=ngg0kplCFxd9Av5IXJBCFtDts0gsiC%2FVbmBj8fXHS8M%3D&reserved=0>, or unsubscribe<https://apc01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFJFEYTSDFV6EBCJPILG5XDRWZLKFANCNFSM4JTB5VTA&data=02%7C01%7Cneha.mahale%40lntebg.com%7Cd9232c8cede2473c6ece08d811521158%7C4852d0fcf87a462bad09773f986ccc04%7C0%7C1%7C637278391470561704&sdata=kAv1WDY%2F4zlo9Qp77Vg0WV0jTLU1gEXtQRMGcNcUv%2BY%3D&reserved=0>.

Larsen & Toubro Limited

www.larsentoubro.com

This Email may contain confidential or privileged information for the intended recipient (s). If you are not the intended recipient, please do not use or disseminate the information, notify the sender and delete it from your system.

**Comment by BarlowHall** [comment]
Sorry for not replying!

I did some tests. I am assuming that the error that the "Save Telemetry" node was giving was
`Can't parse value: {"key1":"value1","key2":"value2"}`. If not, then I will need to know what the error was.

It is because the "values" key is a JSON, which is not a valid value for a time series entry. You can use `JSON.stringify()` to cast it to a string in the rule node, then later in the post processing function you using `JSON.parse()` to change it back to a JSON.

Hope this helps!

**Comment by bigdolphin** [comment]
Hi, I am trying to use your code but it says self is not define. Could you help me? My version is 3.3.4.1.
Thank you very much.

**Comment by BarlowHall** [comment]
Hi,

Sorry for the slow reposne. 

The fix only works on version 2.5.6

Regards

Neale
